spanner/ddl/ast.go

```diff
@@ -20,110 +20,67 @@
 // https://cloud.google.com/spanner/docs/data-definition-language.
 // Before defining each type, we give the snippet from this definition.
 //
-// We use go interface types to preserve the structural constraints of
-// Spanner DDL. For example, suppose ScalarType could be either
-// an INT or a STRING with a length specifier:
-//    ScalarType := INT | STRING( length )
-// then we use a go interface type to encode ScalarType:
-//    type ScalarType interface {
-//        PrintScalarType() string  // To print ScalarTypes types.
-//    }
-// and struct types for each case:
-//    type Int struct { }
-//    type String struct { length int64 }
-// and finally, we define functions:
-//    func (i Int) PrintScalarType() { ... }
-//    func (s String) PrintScalarType() { .. }
-//
-// The net result is that the only way to build a ScalarType
-// is using Int or String.
+// Package ddl provides a generic representation of Spanner DDL.
 package ddl
 
 import (
 	"fmt"
 	"strings"
 )
 
-// Length encodes the following Spanner DDL definition:
-//     length:
-//        { int64_value | MAX }
-type Length interface {
-	PrintLength() string
-}
-
-// Int64Length wraps an integer length specifier.
-type Int64Length struct{ Value int64 }
-
-// MaxLength represents "MAX".
-type MaxLength struct{}
-
-// Interface validation for Length.
-var _ = []Length{Int64Length{}, MaxLength{}}
+const (
+	// Bool represent BOOL scalar type
+	Bool string = "BOOL"
+	// Bytes represent BYTES scalar type
+	Bytes string = "BYTES"
+	// Date represent DATE scalar type
+	Date string = "DATE"
+	// Float64 represent FLOAT64 scalar type
+	Float64 string = "FLOAT64"
+	// Int64 represent INT64 scalar type
+	Int64 string = "INT64"
+	// String represent STRING scalar type
+	String string = "STRING"
+	// Timestamp represent TIMESTAMP scalar type
+	Timestamp string = "TIMESTAMP"
+	// Max represent "MAX" length of scalar type.
+	Max string = "MAX"
+)
 
-// PrintLength unparses Int64Length.
-func (i Int64Length) PrintLength() string { return fmt.Sprintf("%d", i.Value) }
+// MaxLength represents "MAX" length for a scalar_type.
+var MaxLength = Max
 
-// PrintLength unparses MaxLength.
-func (m MaxLength) PrintLength() string { return fmt.Sprintf("MAX") }
+// PrintLength unparses length of a scalar_type.
+func PrintLength(m interface{}) string {
+	return fmt.Sprintf("%v", m)
+}
 
-// ScalarType encodes the following DDL definition:
+// Type represents the scalar_type of a column.
 //     scalar_type:
 //        { BOOL | INT64 | FLOAT64 | STRING( length ) | BYTES( length ) | DATE | TIMESTAMP }
-type ScalarType interface {
-	PrintScalarType() string
+type Type struct {
+	Name string
+	// Len encodes the following Spanner DDL definition:
+	//     length:
+	//        { int64_value | MAX }
+	Len interface{}
```

Comment:
I'm ok with this encoding and I don't have anything better to suggest.
But just wanted to note that, in principle, we have three different things to represent:

a) an integer length
b) a MAX length
c) no length

The interface encoding is kind of convenient since it allows us to write:

  Type{Name: Int64}
  Type{Name: String, Len: MaxLength}
  Type{Name: Bytes, Len: int64(42)}

But it allows us to put anything as a Len, including an arbitrary string.

Reply:
Yes, that's correct. `Type.Len` could be `int64(value)`, `MaxLength` or `NULL` value. But that allows us to put other arbitrary strings as well. This is a trade off for such refactoring.

Reply:
I would recommend you to take a look at this link: https://github.com/googleapis/google-cloud-go/blob/005eb3d360540e4747b607cc681950fa2ebb2d82/spanner/spansql/types.go#L245-L253. 

This is the existing code in `spanner/spansql`. I would prefer to define `Len` as `Int64`.

Reply:
Probably hold on. The only concern of my suggestion is that if we encode it into json, it will become a very large number in the file everywhere when there is a STRING(MAX). The good part is that the Go type is fixed.

Reply:
This isn't too bad.
STRING(MAX) would get encoded as:
{"Name":"STRING","Len":9223372036854775807}
Its isn't likely to cause a huge increase in json for our configs, and the size isn't really an issue in any case, since schemas are pretty small.

Also, I like the idea of consistency with google-cloud-go/spanner/spansql/types.go.
At some point we may want to consider replacing ast.go with those definitions (but that's a discussion for another day).

Reply:
@Dharmil-searce Can you please make the change according to our discussion here? Thanks.

Reply:
Done. I have updated `Len` to `Int64` and tried to keep `PrintColumnDefType` consistent with `spanner/spansql` code.